chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2513 @@
+        rf.valid_tuning_steps = STEPS
+        # rf.valid_dtcs_codes = UV5R_DTCS
+
+        normal_bands = [self._vhf_range, self._uhf_range]
+        # rax_bands = [self._vhf_range, self._220_range]
+
+        if self._mmap is None:
+            rf.valid_bands = [normal_bands[0], normal_bands[1]]
+        # elif not self._is_orig() and self._my_upper_band() == vhf_220_radio:
+        #     rf.valid_bands = rax_bands
+        else:
+            rf.valid_bands = normal_bands
+        rf.memory_bounds = (1, 199)
+        return rf
+        
+        
+    def process_mmap(self):
+        self._memobj = bitwise.parse(MEM_FORMAT, self._mmap)
+
+    def validate_memory(self, mem):
+        msgs = chirp_common.CloneModeRadio.validate_memory(self, mem)
+
+        _msg_duplex2 = 'Memory location only supports "(None)" or "off"'
+        _msg_duplex3 = 'Memory location only supports "(None)", "+" or "off"'
+
+        if self._gmrs:
+            if mem.number < 1 or mem.number > 30:
+                if float(mem.freq) / 1000000 in GMRS_FREQS1:
+                    if mem.duplex not in ['', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex2))
+
+                if float(mem.freq) / 1000000 in GMRS_FREQS2:
+                    if mem.duplex not in ['', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex2))
+
+                if float(mem.freq) / 1000000 in GMRS_FREQS3:
+                    if mem.duplex not in ['', '+', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex3))
+
+        return msgs
+    
+    # def mode_block(self):
+    #     if MODE_RADIO == 'P31185': #ham
+    #         return 1
+    #     else:
+    #         return 
+    
+    def sync_in(self):
+        self.ident_Mode
+        try:
+            self._mmap = _do_download(self)
+        except errors.RadioError:
+            raise
+        except Exception as e:
+            raise errors.RadioError("Failed to communicate with radio: %s" % e)
+        self.process_mmap()
+        
+        # HAM OR GMRS
+        LOG.info("Radio mode is " + MODE_RADIO)
+        LOG.info("Chirp choose mode is " + self.ident_Mode)
+        # The Ham and Gmrs modes are subclasses of this model tdh8. 
+        # When the Ham or Gmrs mode is selected, 
+        # a double verification operation will be performed to 
+        # obtain the current mode code of the walkie-talkie and 
+        # the mode code currently selected by the chirp for comparison. 
+        # If the two are consistent, enter This mode, 
+        # otherwise an inconsistent error will be prompted.
+        
+        if MODE_RADIO == 'P31183' and self.ident_Mode == 'P31183': # normal mode
+            self._memobj.settings.ham = 0
+            self._memobj.settings.gmrs = 0
```

Comment:
ok i will try to follow your way.

Reply:
Cool, the users will appreciate that :)

Reply:
I found that I have to perform a cloning routine to obtain the logo of this model, because the cloning will communicate with the walkie-talkie to obtain the logo. As for the model judgment when selecting, I am still not sure, but I can choose Then prompt the user like this.

Reply:
I think you're misunderstanding me - you're already detecting the model early in get_ident(). I will comment up there.